rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -528,100 +606,172 @@ def slice_state_history(
+            An iterator over state, labels, entity tag tuples
+        """
+
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+
+        label_index = 0
+        entity_data = {}
+        for event in tracker.applied_events():
+            if isinstance(event, UserUttered):
+                entity_data = self._entity_data(event)
+
+            elif isinstance(event, ActionExecuted):
+
+                label_index += 1
 
                 # use only actions which can be predicted at a stories start
                 if event.unpredictable:
                     continue
 
                 sliced_states = self.slice_state_history(
-                    states[:states_length_for_action], self.max_history
+                    tracker_states[:label_index], self.max_history
                 )
-                if self.remove_duplicates:
-                    hashed = self._hash_example(
-                        sliced_states, event.action_name or event.action_text, tracker
-                    )
-
-                    # only continue with tracker_states that created a
-                    # hashed_featurization we haven't observed
-                    if hashed not in hashed_examples:
-                        hashed_examples.add(hashed)
-                        trackers_as_states.append(sliced_states)
-                        trackers_as_actions.append(
-                            [event.action_name or event.action_text]
-                        )
-                        trackers_as_entities.append([entity_data])
-                else:
-                    trackers_as_states.append(sliced_states)
-                    trackers_as_actions.append([event.action_name or event.action_text])
-                    trackers_as_entities.append([entity_data])
+                label = [event.action_name or event.action_text]
+                entities = [entity_data]
+
+                yield sliced_states, label, entities
 
                 # reset entity_data for the the next turn
                 entity_data = {}
-                pbar.set_postfix({"# actions": "{:d}".format(len(trackers_as_actions))})
-
-        self._remove_user_text_if_intent(trackers_as_states)
 
-        logger.debug("Created {} action examples.".format(len(trackers_as_actions)))
+    def _check_example_cache(
+        self, tracker: DialogueStateTracker, states: List[State], label: List[Text],
+    ) -> bool:
+        """Returns True if training example is a duplicate."""
+        if not self.remove_duplicates:
+            return False
+        else:
+            hashed = self._hash_example(tracker, states, label)
+            if hashed not in self.hashed_examples:
+                self.hashed_examples.add(hashed)
+                return False
+            else:
+                return True
 
-        return trackers_as_states, trackers_as_actions, trackers_as_entities
+    def _cleanup_example_iterator(self) -> None:
```

Comment:
as before probably not necessary

Reply:
Removing these setup and cleanup functions.